cedar-policy-core/src/parser.rs

```diff
@@ -268,23 +275,78 @@ pub(crate) fn parse_literal(val: &str) -> Result<ast::Literal, Vec<err::ParseErr
+    let mut errs = err::ParseErrors::new();
     // we need to add quotes for this to be a valid string literal
-    text_to_cst::parse_primary(&format!(r#""{val}""#))?
-        .to_string_literal(&mut errs)
-        .ok_or(errs)
+    let cst = text_to_cst::parse_primary(&format!(r#""{val}""#))?;
+    let Some(ast) = cst.to_string_literal(&mut errs) else { return Err(errs); };
+    if errs.is_empty() {
+        Ok(ast)
+    } else {
+        Err(errs)
+    }
 }
 
 /// parse an identifier
 ///
 /// Private to this crate. Users outside Core should use `Id`'s `FromStr` impl
-/// or its constructors
-pub(crate) fn parse_ident(id: &str) -> Result<ast::Id, Vec<err::ParseError>> {
-    let mut errs = Vec::new();
-    text_to_cst::parse_ident(id)?
-        .to_valid_ident(&mut errs)
-        .ok_or(errs)
+/// or its constructorsVec<err::Error>Vec<err::Error>Vec<err::Error>
+pub(crate) fn parse_ident(id: &str) -> Result<ast::Id, err::ParseErrors> {
+    let mut errs = err::ParseErrors::new();
+    let cst = text_to_cst::parse_ident(id)?;
+    let Some(ast) = cst.to_valid_ident(&mut errs) else { return Err(errs); };
+    if errs.is_empty() {
+        Ok(ast)
+    } else {
+        Err(errs)
+    }
+}
+
+/// parse into a `Request`
+pub fn parse_request(
+    principal: impl AsRef<str>,      // should be a "Type::EID" string
+    action: impl AsRef<str>,         // should be a "Type::EID" string
+    resource: impl AsRef<str>,       // should be a "Type::EID" string
+    context_json: serde_json::Value, // JSON object mapping Strings to ast::RestrictedExpr
+) -> Result<ast::Request, err::ParseErrors> {
+    let mut errs = err::ParseErrors::new();
+    // Parse principal, action, resource
+    let mut parse_par = |s, name| {
+        parse_euid(s)
+            .map_err(|e| {
+                errs.push(err::ParseError::WithContext {
+                    context: format!("trying to parse {}", name),
+                    errs: e,
+                })
+            })
+            .ok()
+    };
+
+    let (principal, action, resource) = (
+        parse_par(principal.as_ref(), "principal"),
+        parse_par(action.as_ref(), "action"),
+        parse_par(resource.as_ref(), "resource"),
+    );
+
+    let context = match ast::Context::from_json_value(context_json) {
+        Ok(ctx) => Some(ctx),
+        Err(e) => {
+            errs.push(err::ParseError::ToAST(format!(
+                "failed to parse context JSON: {}",
+                err::ParseErrors(vec![err::ParseError::ToAST(e.to_string())])
+            )));
+            None
+        }
+    };
+    match (principal, action, resource, errs.as_slice()) {
+        (Some(p), Some(a), Some(r), &[]) => Ok(ast::Request {
+            principal: ast::EntityUIDEntry::concrete(p),
+            action: ast::EntityUIDEntry::concrete(a),
+            resource: ast::EntityUIDEntry::concrete(r),
+            context,
+        }),
+        _ => Err(errs),
+    }
```

Comment:
this was recently removed -- do we need to add it back?

Reply:
Nope, must be an artifact of rebasing from `2.2` -> `2.3`.